lib/api/multiObjectDelete.js

```diff
@@ -191,166 +248,187 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
+                    // is specified, else continue to create a delete marker
+                    if (!versionId || !bucket.isObjectLockEnabled()) {
+                        return callback(null, objMD, versionId);
                     }
-                    if (versionId && objMD.location &&
-                        Array.isArray(objMD.location) && objMD.location[0]) {
-                        // we need this information for data deletes to AWS
-                        // eslint-disable-next-line no-param-reassign
-                        objMD.location[0].deleteVersion = true;
+                    const objLockInfo = new ObjectLockInfo({
+                        mode: objMD.retentionMode,
+                        date: objMD.retentionDate,
+                        legalHold: objMD.legalHold || false,
+                    });
+
+                    // If the object can not be deleted raise an error
+                    if (!objLockInfo.canModifyObject(hasGovernanceBypass)) {
+                        log.debug('trying to delete locked object');
+                        return callback(objectLockedError);
                     }
+
                     return callback(null, objMD, versionId);
-                }),
-            (objMD, versionId, callback) => {
-                // AWS only returns an object lock error if a version id
-                // is specified, else continue to create a delete marker
-                if (!versionId || !bucket.isObjectLockEnabled()) {
-                    return callback(null, null, objMD, versionId);
-                }
-                const hasGovernanceBypass = hasGovernanceBypassHeader(request.headers);
-                if (hasGovernanceBypass && authInfo.isRequesterAnIAMUser()) {
-                    return checkUserGovernanceBypass(request, authInfo, bucket, entry.key, log, error => {
-                        if (error && error.is.AccessDenied) {
-                            log.debug('user does not have BypassGovernanceRetention and object is locked', { error });
-                            return callback(objectLockedError);
+                },
+                (objMD, versionId, callback) => {
+                    const options = preprocessingVersioningDelete(
+                        bucketName, bucket, objMD, versionId, config.nullVersionCompatMode);
+                    const deleteInfo = {};
+                    if (options && options.deleteData) {
+                        deleteInfo.deleted = true;
+                        if ((bucket.getLifecycleConfiguration && !bucket.getLifecycleConfiguration())
+                            && (bucket.getNotificationConfiguration && !bucket.getNotificationConfiguration())) {
+                            options.shouldOnlyDelete = true;
                         }
-                        if (error) {
-                            return callback(error);
+                        if (objMD.uploadId) {
```

Comment:
this condition may change in the future : i find it should be isolated in a function so that we can easily evolve and maintain it (and possibly introduce more complex conditions)

Reply:
I created https://github.com/scality/cloudserver/pull/5195/commits/21d22ad1832687ed188f284dcafeaf6ae1c4f285#diff-b90cf333e787e31130d0434b2e48b75b69948c9e9c594c2d5f201058535b6ef4R1 that is also used when doing single deletes